samples/bootloader/README.rst

```diff
@@ -0,0 +1,170 @@
+
+See :ref:`ug_bootloader` for more information about the full bootloader chain.
+
+Overview
+********
+
+The bootloader sample provides a simple root of trust (RoT) anchor as well as support for an upgradable second stage bootloader.
+
+This is accomplished by the following steps:
+
+1. Lock the flash.
+     To enable the RoT, the bootloader sample locks the flash that contains the sample bootloader and its configuration.
+     Locking is done using the hardware that is available on the given architecture.
+     For details on locking, see the :ref:`fprotect_readme` driver.
+
+#. Select the next stage in the boot chain.
+     When the bootloader sample is enabled, there are two slots in which the second stage bootloader can reside.
+     The second stage bootloader in each slot has a version number associated with it, and the bootloader sample selects the second stage bootloader that has the highest version number.
+
+#. Verify the next stage in the boot chain.
+     After selecting the image to be booted next, the bootloader sample verifies its validity using one of the provisioned public keys hashes.
+     The next boot stage has a set of metadata associated with it.
+     This metadata contains the full public key corresponding to the private key that was used to sign the firmware.
+     To save space, the bootloader sample checks only the hash of this public key against a set of provisioned hashes.
+     If the public key in the metadata matches one of the provisioned public key hashes, the image is considered valid.
+     If the public key does not match any of the provisioned hashes, validation fails.
+
+#. Boot the next stage in the boot chain.
+    After verifying the next boot stage, the bootloader sample uninitializes all peripherals that it used and boots the next boot stage.
+
+#. Share the cryptographic library over ABI.
+     The bootloader sample does not contain any flash writing code.
+     Therefore, it is safe to share some of its functionality through an Application Binary Interface (ABI, see :ref:`fw_info_abi`).
+     For more information, see :file:`bl_crypto.h`.
+
+
+Flash layout
+============
+
+The flash layout is defined by :file:`samples/bootloader/pm.yml`.
+
+The bootloader sample defines four main areas:
+
+1. **B0** - Contains the bootloader sample.
+#. **Provision** - Stores the provisioned data.
+#. **S0** - One of two potential storage areas for the second stage bootloader.
+#. **S1** - One of two potential storage areas for the second stage bootloader.
+
+Provisioning
+============
+
+The public key hashes are not compiled in with the source code of the bootloader sample.
+Instead, they are stored in a separate memory region through a process called *provisioning*.
+This is done to facilitate the manufacturing process of a device with the bootloader sample, as the generation of public key hashes is performed at the same time as the generation of the private keys used for signing.
```

Comment:
The steps are:
1 Compile the bootloader, the bootloader is now not able to run, as it does not see any public keys it can use for verification.
2 Generate a set of private-public keypairs
3 Generate a hex file, in a specific format which contains the hash of the public keys
4 Write (provision) the generated hex file to the device, the bootloader can now use the provisioned keys.

Step 1 can be performed at some huge open factory, while step 2 and onwards is performed in a more secure manner.

By default the build system will generate keys and provision them automatically. These can easily be overwritten using the provided scripts.

Reply:
reworded